caffe2/onnx/backend.cc

```diff
@@ -0,0 +1,1080 @@
+  if (not onnx_tensor.has_raw_data()) {
+    return false;
+  }
+
+  size_t raw_size = onnx_tensor.raw_data().size();
+  assert(raw_size % sizeof(T) == 0);
+
+  size_t num_elements = raw_size / sizeof(T);
+  const void *src_ptr =
+      static_cast<const void *>(onnx_tensor.raw_data().data());
+  field->Resize(num_elements, 0);
+  void *target_ptr = static_cast<void *>(field->mutable_data());
+  memcpy(target_ptr, src_ptr, raw_size);
+
+  return true;
+}
+
+//TODO We probably should have only one copy of this function (copied from pybind_state.cc)
+bool ParseProtobufFromLargeString(const std::string &str, ::google::protobuf::Message *proto) {
+  ::google::protobuf::io::ArrayInputStream input_stream(str.data(), str.size());
+  ::google::protobuf::io::CodedInputStream coded_stream(&input_stream);
+  // Set PlanDef message size limit to 1G.
+  coded_stream.SetTotalBytesLimit(1024LL << 20, 512LL << 20);
+  return proto->ParseFromCodedStream(&coded_stream);
+}
+
+caffe2::DeviceOption GetDeviceOption(const onnx::Device &onnx_device) {
+  static const std::unordered_map<onnx::DeviceType, caffe2::DeviceType> m = {
+    {onnx::DeviceType::CPU, caffe2::DeviceType::CPU},
+    {onnx::DeviceType::CUDA, caffe2::DeviceType::CUDA}
+  };
+  caffe2::DeviceOption d;
+  d.set_device_type(static_cast<int32_t>(m.at(onnx_device.type)));
+  d.set_cuda_gpu_id(onnx_device.device_id);
+  return d;
+}
+
+// Note: if we design the API a bit, we can avoid the two copies
+onnx::ModelProto OptimizeOnnx(const onnx::ModelProto &input, bool init) {
+  std::vector<std::string> passes{"fuse_consecutive_transposes",
+                                  "eliminate_nop_transpose",
+                                  "fuse_transpose_into_gemm"};
+
+  if (init) {
+    passes.emplace_back("split_init");
+  } else {
+    passes.emplace_back("split_predict");
+  }
+  return onnx::optimization::Optimize(input, passes);
+}
+
+template <class T, class U>
+U LookUpWithDefault(const std::unordered_map<T, U> &map, const T &key,
+                    const U &default_value) {
+  const auto it = map.find(key);
+  if (it == map.end()) {
+    return default_value;
+  } else {
+    return it->second;
+  }
+}
+
+const onnx::OpSchema &GetOnnxSchema(const std::string &key) {
+  const auto *schema = onnx::OpSchemaRegistry::Schema(key);
+  if (!schema) {
+    throw std::runtime_error("No ONNX schema registered for '" + key + "'!");
+  }
+  return *schema;
+}
+
+void BuildOperator(
+    caffe2::OperatorDef* c2_op,
+    const std::string &op_type,
+    const std::vector<std::string> &inputs,
+    const std::vector<std::string> &outputs,
+    const std::vector<caffe2::Argument>& args) {
+  c2_op->set_name("");
+  c2_op->set_type(op_type);
+  for (const auto& input: inputs) {
+    c2_op->add_input(input);
+  }
+  for (const auto& output: outputs) {
+    c2_op->add_output(output);
+  }
+  for (const auto& arg: args) {
+    auto* tmp = c2_op->add_arg();
+    assert(tmp);
```

Comment:
it should be always not null.

also we usually avoid asserts as they're compiled away anyway. Thus CHECK for hard invariants (rare) or CAFFE_ENFORCE for exceptions